Implement File block

mirka opened the pull request:
Closes #5462

## Description
Implements a File block with the following features:
- [x] Editable text link
- [x] Text link can be copy and pasted to create a normal link
- [x] HTML5 download link
- [x] Copy URL button to copy the media file URL to clipboard
- [x] Accepts ~~both Media Library files and external URLs~~ only Media Library files
- [x] Text link can be changed to point to the Attachment Page
- [x] Text link can be changed to open in a new window
- [x] Accepts and uploads a file drag-and-dropped into the editor (or placeholder)
- [x] Transforms to/from Audio and Video blocks
- [x] Download button text is editable
- [x] Download button is optional
- ~~(?) Background Color and Text Color options in sidebar~~

## How has this been tested?
Try the features above.

## Screenshots <!-- if applicable -->
<img width="502" alt="screen shot 2018-05-27 at 5 56 09" src="https://user-images.githubusercontent.com/555336/40580102-ab6f7fa2-6172-11e8-8ddf-d0a8d4f91e18.png">

## Things I don't plan on including in this PR (but should be done eventually)
- ~~Integrate error messaging using #4203~~ done, thanks to #6820
- ~~Add allowed mime type checking to #4203~~ submitted separately as #6968
- ~~Refactor the misplaced CSS in [core-blocks/image/editor.scss](https://github.com/WordPress/gutenberg/pull/6805/files#diff-1e7ab348c11a3ea0cec34ff2eb2d8337L62)~~ done, thanks to #6820
- Add drag and drop support to Audio/Video blocks so they can take precedence over this generic File block
- ~~Copy URL button may not work correctly in Safari~~ submitted separately as #7106

mirka committed:
Add basic file block

mirka committed:
Reset id attribute when new URL is specified

mirka committed:
Remove extra margin on file block placeholder

mirka committed:
Upload files dropped into editor

mirka committed:
Add transient animation while uploading file

mirka committed:
Don’t call string method on undefined url

mirka committed:
Download file with filename

mirka committed:
Add option to link to attachment page

mirka committed:
Only disassociate id when URL was changed

mirka committed:
Split out file block inspector controls

mirka committed:
Add option to open in new window

mirka committed:
Make filename editable

mirka committed:
Add hover style to button

mirka committed:
Only render when href is not empty

mirka committed:
Add noreferrer noopener to target _blank links

mirka committed:
Make text links copy-pasteable

mirka committed:
Allow empty filenames

mirka committed:
Add DropZone to file block

mirka committed:
Make file block transformable to/from audio/video

mirka committed:
Fix cursor disappearing while editing text link

mirka committed:
Filter out premature onChange events in CJK

mirka committed:
Tweak code readability

mirka committed:
Fix line break filtering

mirka committed:
Remove unnecessary RichText component in save()

mirka committed:
Add tests

jasmussen commented:
Tried to check this out and run it. VERY cool work! Very very nice, definitely getting there.

A few niggles:

Can we make it so you can drag and drop a PDF file directly into the canvas, in addition to on the file block placeholder?

It would be nice if the Download button text were editable. Perhaps even that you could toggle off the download button entirely in the sidebar. If you remove the actual link from the button inside the editor, so you can't download from inside the editor but only on the published front-end, then you could make the button clickable to edit, just like the normal Button block.

Given that you can edit the text of the link, right now the "Edit" button taking you back to the placeholder feels a little bit jarring. Could you make it so instead of having the layout alignment buttons and an edit button, it has instead text-block-like editing tools?

Instead of:

<img width="333" alt="screen shot 2018-05-18 at 09 03 34" src="https://user-images.githubusercontent.com/1204802/40220738-7ff8ed94-5a7a-11e8-8212-7ce115e132c0.png">

You'd have:

<img width="326" alt="screen shot 2018-05-18 at 09 03 30" src="https://user-images.githubusercontent.com/1204802/40220739-8413a3ce-5a7a-11e8-9ce9-f70f549da401.png">

This would essentially be like a text field with a button. You could write your own text, link, unlink, bold, italic, etc. You'd never be able to go back to the placeholder block from the final state, but that's fine too, and the button would always hold the link. What do you think?

I notice by the way that if you edit the link, you're typing in reverse :) — but I'll bet that's what you refer to in the remaining bullet.

![reverse](https://user-images.githubusercontent.com/1204802/40220831-d02616de-5a7a-11e8-8211-6bf0544d2e8a.gif)

> Background Color and Text Color options in sidebar

We probably don't need these, but nice idea otherwise!

Very cool work, thanks for working on this, and keep up he good work!

mirka commented:
Thanks for the review, @jasmussen!

> Can we make it so you can drag and drop a PDF file directly into the canvas, in addition to on the file block placeholder?

Do you mean like this?
![direct-pdf-drop](https://user-images.githubusercontent.com/555336/40244281-6b9e852e-5afd-11e8-8bd3-a24a6d06b773.gif)

If so, it is already supposed to be working. You might be seeing a bug. Could you give me some more details so I can reproduce it?

(BTW are you rebasing onto a branch that is not `master`? It looks like you have icons <img width="68" alt="screen shot 2018-05-18 at 23 02 20" src="https://user-images.githubusercontent.com/555336/40244330-8a57db1e-5afd-11e8-9ab0-51b64e0b2f8b.png"> in the toolbar that I don't)

> It would be nice if the Download button text were editable. Perhaps even that you could toggle off the download button entirely in the sidebar.

Sure! We could do that.

> This would essentially be like a text field with a button. You could write your own text, link, unlink, bold, italic, etc. You'd never be able to go back to the placeholder block from the final state, but that's fine too, and the button would always hold the link. What do you think?

Hmm, I thought about this too. At one point I even considered a hybrid approach:

![artboard](https://user-images.githubusercontent.com/555336/40244420-ccebd8d6-5afd-11e8-9052-b1fc5f98c7ba.png)
(I scrapped this idea, because having a link icon there would make the pencil icon a lot less discoverable)

My main concern is that we’d confuse the user’s mental model of where the “source of truth” for the file link is.

Since choosing a file from the placeholder sets both links, you can reasonably assume that both links will change if you choose another file from the same interface:

![artboard 3](https://user-images.githubusercontent.com/555336/40244534-1360cf60-5afe-11e8-84f3-de4e495f64a3.png)

But when you introduce a second way to change the link, it’s harder to tell what’s going to happen. If only the text link were to change, it would become hard to tell that the text link and button link are out of sync. On the other hand, if both links were to change, it would break the expectation of the standard rich text UI:

![artboard 2](https://user-images.githubusercontent.com/555336/40244543-17a9967e-5afe-11e8-8420-e1869bff9289.png)

Is this a tradeoff we are willing to take? Let me know what you think 🙂

SuperGeniusZeb commented:
> (BTW are you rebasing onto a branch that is not master? It looks like you have icons in the toolbar that I don't)

Those are the wide and full-width alignment options, which do not show unless your theme declares support for them, since those kinds of alignments break out of the standard content width and may not be desired depending on the theme. Since Gutenberg is so new, there are very few themes that have declared support for those alignment types, so most of the time you will only see those options if you use a custom child theme which has declared support for them.

mirka commented:
@SuperGeniusZeb Oh I see, that makes sense now. Thank you!

mirka committed:
Merge branch 'master' of https://github.com/WordPress/gutenberg into 6805

mirka committed:
Make download button optional

mirka committed:
Make button text editable

mirka committed:
Merge branch 'master' of https://github.com/WordPress/gutenberg into 6805

# Conflicts:
#	core-blocks/image/editor.scss
#	utils/mediaupload.js

jasmussen commented:
My apologies, @mirka, for not responding sooner! Busy days these! Please feel free to ping me on Slack (@jasmussen) if I should forget a review another time.

> If so, it is already supposed to be working. You might be seeing a bug. Could you give me some more details so I can reproduce it?

Yes, exactly so. I don't know why it didn't work when I tried it before. A cache ghost perhaps? In any case, it's working exactly as it should now. Nice work!

> Hmm, I thought about this too. At one point I even considered a hybrid approach:
> (I scrapped this idea, because having a link icon there would make the pencil icon a lot less discoverable)
> My main concern is that we’d confuse the user’s mental model of where the “source of truth” for the file link is.
> Since choosing a file from the placeholder sets both links, you can reasonably assume that both links will change if you choose another file from the same interface:

This is an excellent thought process, and touches on probably the key challenge of this block. Is it a file block? Or is it a text block that has a download button and conveniences?

Tricky tricky challenge. I think we should try a couple of things. 

1. First off, let's make it so that the placeholder block is only ever a placeholder block, and you can never "return" to that. Right now when you click the Edit button you return the block to a placeholder block, this breaks the intent of the placeholder for the sake of showing a URL bar. This acknowledges the work that was just merged in https://github.com/WordPress/gutenberg/pull/6820, which unifies all the media placeholder blocks into one pattern. 

2. Incidentally that means we should remove the "URL" aspect from the placeholder block entirely. This:

<img width="461" alt="screen shot 2018-05-24 at 09 48 14" src="https://user-images.githubusercontent.com/1204802/40471646-9ca37056-5f37-11e8-9b90-eb30cc70b1be.png">

That means for this PR, you won't be able to add a file from a URL at all. This is fine for the scope, especially since that is tracked separately in #1536. And since we have a new unified media block, then once we decide on a UI & flow for how to add media from a URL, presumably we can roll that feature out to all media blocks, including this one.

3. We should then make it, as is already the case and as you suggest, where there is a single source of truth for the URL of the image is the media library. Let's mimic what the Image block does here:

![media](https://user-images.githubusercontent.com/1204802/40471809-0d7410b0-5f38-11e8-9739-a3672b1e8200.gif)

That is — once the edit button takes you to the media library. You can here either upload a new file and choose that, or you can simply pick a different file to link to. Both the Download button and the URL of the text are updated.

4. We should add a "Copy URL" button next to the download button. If you like, you can make an option to toggle that off in the inspector, but that's a stretch goal. This is important because then if a user doesn't like how the file block looks, or simply wants to link the file itself in text, they can then simply copy the URL, delete the file block, and paste the link in a paragraph.

By the way — it's already working that you can select the text, copy and paste it in a paragraph link, and the link is transformed into an editable one. Very nice! ⭐️ 

5. Can we hide the link boundaries from the file block? They imply the link is editable in text, which it won't be once 3 is addressed.

The link boundary is the light blue background of the link:

<img width="186" alt="screen shot 2018-05-24 at 09 54 33" src="https://user-images.githubusercontent.com/1204802/40471971-7ed88902-5f38-11e8-844c-e0afeeda2bfe.png">

I.e. it should look like the following, even if your cursor is in the text:

<img width="225" alt="screen shot 2018-05-24 at 09 54 58" src="https://user-images.githubusercontent.com/1204802/40472006-9394c284-5f38-11e8-9bc4-de709a02315a.png">

Thank you so much for the work, very cool, and please ping me again if you have questions!

mirka committed:
Merge branch 'master' of https://github.com/WordPress/gutenberg into 6805

mirka commented:
Thanks @jasmussen! I completely agree with your solution, to mimic the Image block flow. Going back to that placeholder is jarring, especially for people who just uploaded a file by drag-and-drop. A lot of those people would think "Why is there a URL here?", and it's unintuitive that the only way to cancel back out of that placeholder is to click the "Use URL" button. (The Audio/Video block placeholders still have this problem.) Good to see it's being worked on in a unified way.

> We should add a "Copy URL" button next to the download button. If you like, you can make an option to toggle that off in the inspector, but that's a stretch goal. This is important because then if a user doesn't like how the file block looks, or simply wants to link the file itself in text, they can then simply copy the URL, delete the file block, and paste the link in a paragraph.

If that is the primary use case, I think it would be best to make the "Copy URL" button appear only in the editor, and only when the block is currently selected. It won't be included in the preview or the actual published post. Does that sound ok?

![artboard 3](https://user-images.githubusercontent.com/555336/40500617-42fe285a-5fc0-11e8-905a-19e41641b77b.png)

![artboard 4](https://user-images.githubusercontent.com/555336/40500626-4c05a806-5fc0-11e8-912f-2e7bb3cee66c.png)

I'll start working on the changes!

mirka committed:
Replace with unified media placeholder

Merges in the changes made in #6820

mirka committed:
Simplify text link editing

mirka committed:
Fix drag-and-drop uploading

Accommodates the changes made to editorMediaUpload() arguments in #4203

mirka committed:
Clean up CSS

mirka committed:
Add Copy URL button

mirka committed:
Fix placeholder not appearing when backspaced

mirka committed:
Make text link copy to clipboard as rich text

mirka committed:
Use custom text link component instead of RichText

Fixes inability to intercept rich text pastes

mirka committed:
Disable multiline editing for download button

mirka committed:
Clarify prop passing to inspector

mirka committed:
Delegate confirm copy timeouts to ClipboardButton

mirka committed:
Fix overridden Copy URL in Safari

Workaround for inconsistent behavior in Safari, where <textarea> is not the document.activeElement at the moment when the copy event fires. This causes documentHasSelection() in the copy-handler component to mistakenly override the ClipboardButton, and copy a serialized string of the current block instead.

mirka committed:
Merge branch 'master' of https://github.com/WordPress/gutenberg into 6805

mirka committed:
Update tests

mirka committed:
Fix placeholder text toggling bug

Placeholder text state was not properly updating when uploading a file from the media placeholder.

jasmussen commented:
Very nice work! Here's a GIF of it:

![file block](https://user-images.githubusercontent.com/1204802/40602799-312a2254-625a-11e8-95dc-a7eb3edeb8aa.gif)

Experience wise, this seems to behave as we've discussed, and given the paradigms Gutenberg works with, this seems like it solves the task at hand:

- allows file uploads
- allows inserting files easily
- allows copying the link so you can use it to create a more regular in-text link

For the scope of this task, 👍 👍, I think this is cool! Nice work.

Did you include/merge the work from master that unifies the placeholder block?

---

At some point in the future, as a separate project to this, I imagine that we'll want to take a look at the media library in WordPress in general. In doing so, there will likely also be an opportunity to look at how the placeholder block can work. For example it would be nice if we could show part of the media library itself inside the placeholder block, and allow you to simply drag into it there, and then insert as file block or insert as plaintext link. But that's something to consider for the future.

mirka commented:
@jasmussen Thank you for all your guidance, and the cool gif! There are so many possibilities with the Gutenberg framework, and I'm excited to see how the UX will evolve. Hope I'll have a chance to contribute more.

> Did you include/merge the work from master that unifies the placeholder block?

Yes. The code is a lot cleaner thanks to that. And there's another modification in the works (#6957), so if that gets merged in first I'll get to delete a few more lines 🙂

nb commented on components/clipboard-button/index.js:
```diff
@@ -83,8 +83,17 @@ class ClipboardButton extends Component {
+		// the document.activeElement at the moment when the copy event fires.
+		// This causes documentHasSelection() in the copy-handler component to
+		// mistakenly override the ClipboardButton, and copy a serialized string
+		// of the current block instead.
```
Thanks for the detailed comment, explaining why do we have the workaround.

nb replied:
One problem in Safari, though, is that the while block (`<!-- p:file {"href":…`) ends up in the clipboard.

mirka replied:
Ok, it looks like this workaround hadn't fixed the Safari bug completely. I did some testing and narrowed down the repro condition a bit. My workaround seems to be effective for pre-existing posts (i.e. posts that were already saved when the page loaded), but not for new posts. Not sure what's going on, but that's a start... I'll have to look into it further.

mirka replied:
Addressed in #7106

nb commented on core-blocks/file/edit.js:
```diff
@@ -0,0 +1,252 @@
+			getBlobByURL( href )
+				.then( ( file ) => {
+					editorMediaUpload( {
+						allowedType: '*',
```
Just to make sure I understand correctly, once #6968 gets in we will feed the list if types here?

gziolo replied:
From #5462:
> WordPress also supports a number of filetypes that aren't media, such as pdf, doc, and others. See full list here: https://codex.wordpress.org/Uploading_Files

I don't think that `editorMediaUpload` in the current shape is what we really want to use in here, because non-media files can be uploaded, too. It seems like we need to refactor the existing method or add a more generic one which `editorMediaUpload` uses behind the scenes.

mirka replied:
Hmm, I was under the impression that the "media" in `editorMediaUpload` and `mediaUpload` is being used in the broader sense of the word, and not to mean "images, audio, and video". This is in line with how WordPress uses the term — every kind of uploadable file, including pdf and other documents, are uploaded to the "Media Library", and are retrieved via the `media` endpoint.

Even [/edit-post/hooks/blocks/media-upload](https://github.com/WordPress/gutenberg/blob/master/edit-post/hooks/blocks/media-upload/index.js), which includes specific logic for when gallery == true, seems generic enough for any kind of file. Is there a reason we want to keep `editorMediaUpload` specifically for image/audio/video files?

mirka replied:
I posted a separate issue for the `allowedType` problem: #7155

nb commented on core-blocks/file/edit.js:
```diff
@@ -0,0 +1,252 @@
+		}
+	}
+
+	isBlobURL( url = '' ) {
```
Would it make sense to move this one to `utils`, like the other blob functions? At which point would you split out `utils/blob`?

mirka replied:
Yes, I think so. But in a separate PR, because I think I saw a couple of other places in the codebase where we’re doing blob url detection, and I’d want to change them all at once.

nb commented on core-blocks/file/edit.js:
```diff
@@ -0,0 +1,252 @@
+
+		// edit component has its own attributes in the state so it can be edited
+		// without setting the actual values outside of the edit UI
+		this.state = {
```
I am not sure I follow the reason we keep the attributes in the state, too. What do we gain?

nb commented on core-blocks/file/edit.js:
```diff
@@ -0,0 +1,252 @@
+
+		if ( id !== undefined ) {
+			// Get Attachment Page URL
+			wp.apiRequest( {
```
Any reason we’re not using [withAPIData](https://github.com/WordPress/gutenberg/tree/master/components/higher-order/with-api-data) here?

mirka replied:
I didn't realize we had that! Cool. I put that in place, however it turns out that the snapshot test helper for our components (`blockEditRender()`) isn't equipped to inject `context` in order for it to work. I've made a patch for it and I'll propose it in a separate PR, but not sure if my implementation is good. Anyway, I'm going to remove the snapshot test for this temporarily.

nb commented on core-blocks/file/edit.js:
```diff
@@ -0,0 +1,252 @@
+			} )
+				.then(
+					( result ) => {
+						this.setState( { attachmentPage: result.link } );
```
Would the `setState` work correctly if the component hasn't been rendered yet?

mirka replied:
Good point. Should've been in `componentDidMount()`.

nb commented on core-blocks/file/edit.js:
```diff
@@ -0,0 +1,252 @@
+
+	componentDidUpdate( prevProps ) {
+		// Reset copy confirmation state when block is deselected
+		if ( prevProps.isSelected && ! this.props.isSelected ) {
```
This worked great for me as a user.

nb commented on core-blocks/file/editable-link.js:
```diff
@@ -0,0 +1,81 @@
+
+	componentDidUpdate( prevProps ) {
+		if ( prevProps.text !== this.props.text ) {
+			this.setState( { showPlaceholder: ! this.props.text } );
```
What are the trade-offs of using state here vs. just using the `text` prop for rendering logic?

mirka replied:
Since I'm resorting to use a content-editable here, this is not a "controlled" component and therefore the `text` prop isn't updated on each keystroke.

mirka commented on core-blocks/file/index.js:
```diff
@@ -0,0 +1,164 @@
+					<a
+						href={ textLinkHref }
+						target={ openInNewWindow }
+						rel={ openInNewWindow ? 'noreferrer noopener' : false }
```
I think it’s because `noopener` doesn’t work in IE/Edge. (see also: https://github.com/WordPress/gutenberg/issues/6186)

mirka committed:
Remove unnecessary state

mirka committed:
Use withAPIData HOC

mirka committed:
Remove snapshot testing for now

Requires a modification in the `blockEditRender()` test helper to inject context, or else the withAPIData HOC won’t work.

mirka committed:
Merge branch 'master' of https://github.com/WordPress/gutenberg into file-block

# Conflicts:
#	utils/mediaupload.js

nb commented on core-blocks/file/edit.js:
```diff
@@ -96,10 +79,7 @@ class FileEdit extends Component {
-			this.setState( {
-				attachmentPage: media.link,
-			} );
+			}, this.props.media.get /* refetch attachment page url */ );
```
Does `setAttributes` accept a second argument? When is `this.props.media.get` called?

mirka replied:
Totally assumed `setAttributes` took the same arguments as `setState`! I meant it as a callback. Thank you.

gziolo commented on core-blocks/file/edit.js:
```diff
@@ -0,0 +1,228 @@
+
+export default compose( [
+	withAPIData( ( props ) => ( {
+		media: `/wp/v2/media/${ props.attributes.id }`,
```
You should be rather using [getMedia](https://github.com/WordPress/gutenberg/search?q=getMedia&unscoped_q=getMedia) selector instead of `withAPIData` which is going to be deprecated as soon as we find a way to port all existing code that handles REST API requests. This will also remove the need to introducing changes proposed #7108.

gziolo replied:
```js
export default compose( [
    withSelect( ( select, props ) => {
        const { getMedia } = select( 'core' );
        return {
            image: getMedia( props.attributes.id ), // not sure if getMedia handles undefined values
        };
    } ),
    withNotices,
] )( FileEdit );
```

mirka replied:
Thank you, I wasn't aware of this. I'll look into it.

gziolo commented on utils/mediaupload.js:
```diff
@@ -33,7 +33,9 @@ export function mediaUpload( {
 	};
-	const isAllowedType = ( fileType ) => startsWith( fileType, `${ allowedType }/` );
+	const isAllowedType = ( fileType ) => {
+		return ( allowedType === '*' ) || startsWith( fileType, `${ allowedType }/` );
```
Should it rather list all allowed file types by allowing an array format?

mirka replied:
Yes, I think it should accept arrays, as well as have a more predictable syntax. More on that here: #7155

I opted for a wildcard here as a stop-gap measure, or else we'd have something like `['audio', 'application', 'video', 'text', 'image', 'font']` which I think is more error-prone for consumers.

mirka committed:
Use withSelect instead of withAPIData

mirka commented:
By the way, here are some of the RichText issues that caused me to resort to a `contenteditable` for the [text link](https://github.com/WordPress/gutenberg/pull/6805/files#diff-4ad8d79ab5bdc1b32f9689b57a731a26) part: #7575, #5657, #7591.

Now that I've sorted through the repro conditions and workarounds for them, I feel like #7591 is the only "blocker" for using a RichText here. If that is addressed, I think we could safely replace the contenteditable.

noisysocks commented:
Thanks for your excellent work on this @mirka. @talldan and I are going to pick up on this in https://github.com/WordPress/gutenberg/pull/7622.
